### ISSUE TITLE
tdnf repo scanning is very slow with large update repositories

### ISSUE THREAD

**Issue by haata** [vmware/tdnf#263]
Why is tdnf so slow with large update repos? And is there anything that can be done such as multithreading or parallel downloads?

2m5.543s (Photon 3) vs. 0m10.092s (Photon 4) - 2021-09-30

Photon 3.0 - 65d8e9399b68
```bash
docker run -it photon:3.0
root [ / ]# time tdnf update -y
Refreshing metadata for: 'VMware Photon Extras 3.0 (x86_64)'
Refreshing metadata for: 'VMware Photon Linux 3.0 (x86_64)'
Refreshing metadata for: 'VMware Photon Linux 3.0 (x86_64) Updates'
photon-updates                           29109   100%
Upgrading:
tdnf-cli-libs                   x86_64          3.1.4-2.ph3           photon-updates   70.19k 71872
tdnf                            x86_64          3.1.4-2.ph3           photon-updates  309.20k 316619
rpm-libs                        x86_64          4.14.3-1.ph3          photon-updates  937.03k 959523
openssl                         x86_64          1.0.2za-1.ph3         photon-updates    4.53M 4749385
libsolv                         x86_64          0.6.35-7.ph3          photon-updates    1.10M 1153396
glibc                           x86_64          2.28-17.ph3           photon-updates    8.79M 9220935
curl-libs                       x86_64          7.78.0-2.ph3          photon-updates  604.77k 619288
curl                            x86_64          7.78.0-2.ph3          photon-updates  263.59k 269921

Total installed size:  16.56M 17360939

Downloading:
tdnf-cli-libs                            35853   100%
tdnf                                    118949   100%
rpm-libs                                366196   100%
openssl                                2189954   100%
libsolv                                 378941   100%
glibc                                  3985243   100%
curl-libs                               305229   100%
curl                                    148743   100%
Testing transaction
Running transaction
Installing/Updating: glibc-2.28-17.ph3.x86_64
Installing/Updating: openssl-1.0.2za-1.ph3.x86_64
Installing/Updating: curl-libs-7.78.0-2.ph3.x86_64
Installing/Updating: tdnf-cli-libs-3.1.4-2.ph3.x86_64
Installing/Updating: rpm-libs-4.14.3-1.ph3.x86_64
Installing/Updating: libsolv-0.6.35-7.ph3.x86_64
Installing/Updating: tdnf-3.1.4-2.ph3.x86_64
Installing/Updating: curl-7.78.0-2.ph3.x86_64
Removing: tdnf-3.1.4-1.ph3.x86_64
Removing: rpm-libs-4.14.2-14.ph3.x86_64
Removing: curl-7.78.0-1.ph3.x86_64
Removing: curl-libs-7.78.0-1.ph3.x86_64
Removing: openssl-1.0.2y-1.ph3.x86_64
Removing: libsolv-0.6.35-6.ph3.x86_64
Removing: tdnf-cli-libs-3.1.4-1.ph3.x86_64
Removing: glibc-2.28-15.ph3.x86_64

Complete!

real	2m5.543s
user	1m53.171s
sys	0m1.059s
root [ / ]# tdnf --version
tdnf: 3.1.4
```

Photon 4.0 - dd07f9531eae
```bash
docker run -it photon:4.0 
root [ / ]# time tdnf update -y
Refreshing metadata for: 'VMware Photon Linux 4.0 (x86_64)'
Refreshing metadata for: 'VMware Photon Extras 4.0 (x86_64)'
Refreshing metadata for: 'VMware Photon Linux 4.0 (x86_64) Updates'
photon-updates                            3238   100%
Upgrading:
krb5                            x86_64          1.17-5.ph4            photon-updates    3.34M 3503386
glibc                           x86_64          2.32-10.ph4           photon-updates    8.72M 9142764
elfutils-libelf                 x86_64          0.181-4.ph4           photon-updates  105.11k 107631
curl-libs                       x86_64          7.78.0-2.ph4          photon-updates  620.55k 635440
curl                            x86_64          7.78.0-2.ph4          photon-updates  263.60k 269924
ca-certificates-pki             x86_64          20201001-2.ph4        photon-updates  766.71k 785110
ca-certificates                 x86_64          20201001-2.ph4        photon-updates  761.37k 779645
bzip2-libs                      x86_64          1.0.8-4.ph4           photon-updates   74.37k 76158

Total installed size:  14.59M 15300058

Downloading:
krb5                                   1222263   100%
glibc                                  2770690   100%
elfutils-libelf                          77497   100%
curl-libs                               283842   100%
curl                                    143789   100%
ca-certificates-pki                     301702   100%
ca-certificates                         323454   100%
bzip2-libs                               39288   100%
Testing transaction
Running transaction
Installing/Updating: glibc-2.32-10.ph4.x86_64
Installing/Updating: krb5-1.17-5.ph4.x86_64
Installing/Updating: ca-certificates-pki-20201001-2.ph4.x86_64
Installing/Updating: curl-libs-7.78.0-2.ph4.x86_64
Installing/Updating: ca-certificates-20201001-2.ph4.x86_64
Installing/Updating: curl-7.78.0-2.ph4.x86_64
Installing/Updating: elfutils-libelf-0.181-4.ph4.x86_64
Installing/Updating: bzip2-libs-1.0.8-4.ph4.x86_64
Removing: curl-7.78.0-1.ph4.x86_64
Removing: curl-libs-7.78.0-1.ph4.x86_64
Removing: krb5-1.17-4.ph4.x86_64
Removing: elfutils-libelf-0.181-3.ph4.x86_64
Removing: bzip2-libs-1.0.8-3.ph4.x86_64
Removing: ca-certificates-20201001-1.ph4.x86_64
Removing: ca-certificates-pki-20201001-1.ph4.x86_64
Removing: glibc-2.32-9.ph4.x86_64

Complete!

real	0m10.092s
user	0m3.906s
sys	0m0.463s
root [ / ]# tdnf --version
tdnf: 3.1.4
```

**Comment by oliverkurth** [comment]
As discussed in the company slack, this is because we keep multiple versions in the photon-updates repo for Photon 3.0, while for Photon 4.0 this only has the latest versions. So the updates repo for Photon 3.0 is much bigger and therefore requires longer download times and processing.

**Comment by oliverkurth** [comment]
We are planning to do something similar to the Photon 3.0 repo to speed it up.

**Comment by oliverkurth** [comment]
PR #293 would address this too.

**Comment by oliverkurth** [comment]
This is now fixed in Photon 3.0 with tdnf 3.1.8 and photon-repos 3.0-6.
